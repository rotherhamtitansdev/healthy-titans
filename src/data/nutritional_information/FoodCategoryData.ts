import { MenuCardProps } from "../../models/MenuCardProps";

const FoodCategoryData: MenuCardProps[] = [
  {
    key: 0,
    name: "Meat",
    path: "Meat",
    firebaseName: "FoodCategories/Meat/Meat.svg",
  },
  {
    key: 1,
    name: "Fish",
    path: "Fish",
    firebaseName: "FoodCategories/Fish/Fish.svg",
  },
  {
    key: 2,
    name: "Fruit & Vegetables",
    path: "FruitAndVegetables",
    firebaseName: "FoodCategories/Fruits/FruitAndVeg.svg",
  },
  {
    key: 3,
    name: "Dairy & Eggs",
    path: "DairyAndEggs",
    firebaseName: "FoodCategories/DairyAndEggs/DairyAndEggs.svg",
  },
  {
    key: 4,
    name: "Fast Food",
    path: "FastFood",
    firebaseName: "FoodCategories/JunkFood/FastFood.svg",
  },
  {
    key: 5,
    name: "Sweets",
    path: "Sweets",
    firebaseName: "FoodCategories/Sweets/Sweets.svg",
  },
];

export default FoodCategoryData;
