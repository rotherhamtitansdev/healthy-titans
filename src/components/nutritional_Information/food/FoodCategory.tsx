import React, { useEffect, useState } from "react";
import { useLocation, useParams, useNavigate } from "react-router";
import FoodCategoryData from "../../../data/nutritional_information/FoodCategoryData";
import FoodSubCategories from "../../../data/nutritional_information/FoodSubCategories";
import { MenuCardProps } from "../../../models/MenuCardProps";
import { MenuTitleProps } from "../../../models/MenuTitleProps";
import Menu from "../../shared/Menu";
import MenuHeader from "../../app_header/header/MenuHeader";
import AppHeader from "../../app_header/AppHeader";

const FoodCategory = () => {
  let { foodCategory } = useParams();
  const location = useLocation();
  const navigate = useNavigate();
  const [getCategoryTitle, setCategoryTitle] = useState("");

  useEffect(() => {
    // find the category from hard coded data and match the path with the current category
    let category = FoodCategoryData.find((element) => element.path === foodCategory)?.name;
    // if the category is not found, search the subcategories for the category
    if (!category) {
      category = FoodSubCategories.find((element) => element.category === foodCategory)?.category;
    }
    // if the category is found then set the category title to the given category
    if (category) {
      setCategoryTitle(category);
    } else {
      // if the category is not found then redirect to the nutritional information page
      navigate("/NutritionalInformation");
    }
  }, []);

  function getFoodData() {
    // If params isn't passed from router (e.g for FruitAndVegetables), get path from URL
    if (!foodCategory) {
      foodCategory = location.pathname.slice(
        location.pathname.lastIndexOf("/") + 1,
        location.pathname.length
      );
    }

    const category = FoodSubCategories.find((element) => element.category === foodCategory);
    return category ? category.options : FoodCategoryData;
  }

  function getTitle(): MenuTitleProps {
    return getCategoryTitle
      ? {
          title: getCategoryTitle,
        }
      : {
          title: "Explore food families",
          subtitle: "Click on a picture to learn more",
        };
  }

  const foodData: MenuCardProps[] = getFoodData();

  return (
    <div>
      <AppHeader>
        <MenuHeader
          title="Food & Nutrition"
          body="Food is important, we all need food to be strong and healthy. From here we can learn about the nutritional values in the foods we eat. This will show you the different types of goodness we can get from the different types of food. We will be able to explore what we need to have a healthy, well-balanced diet including carbohydrates, protein as well as vitamins and minerals."
        />
      </AppHeader>
      <Menu title={getTitle()} cards={foodData} />
    </div>
  );
};

export default FoodCategory;
