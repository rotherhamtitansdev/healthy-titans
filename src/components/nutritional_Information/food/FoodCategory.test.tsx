import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { Route } from "react-router";
import userEvent from "@testing-library/user-event";
import FoodCategory from "./FoodCategory";
import RoutingTestWrapper from "../../../tests/RoutingTestWrapper";
import FoodCategoryData from "../../../data/nutritional_information/FoodCategoryData";

describe("Food Category Component", () => {
  test("renders with no category", async () => {
    const route = "/NutritionalInformation";

    const { asFragment } = render(
      <RoutingTestWrapper path={route}>
        <Route path={route} element={<FoodCategory />} />
      </RoutingTestWrapper>
    );

    expect(screen.getByText(/explore food families/i)).toBeVisible();
    expect(screen.getByText(/click on a picture to learn more/i)).toBeVisible();

    expect(asFragment()).toMatchSnapshot();
  });

  test("renders with category", () => {
    const category = FoodCategoryData[0];
    const route = `/NutritionalInformation/${category.name}`;

    render(
      <RoutingTestWrapper path={route}>
        <Route path={route} element={<FoodCategory />} />
      </RoutingTestWrapper>
    );

    expect(screen.getByTestId("menutitle-title")).toHaveTextContent(category.name);
  });

  test("fallsback if no category is found", () => {
    const route = "/NutritionalInformation/NotARealCategory";

    render(
      <RoutingTestWrapper path={route}>
        <Route path={route} element={<FoodCategory />} />
      </RoutingTestWrapper>
    );

    waitFor(() => {
      expect(window.location.pathname).toBe("/NutritionalInformation");
    });
  });

  test("renders with no category & can click on a card", async () => {
    const category = FoodCategoryData[0];
    const route = "/NutritionalInformation";
    const user = userEvent.setup();

    render(
      <RoutingTestWrapper path={route}>
        <Route path={route} element={<FoodCategory />} />
      </RoutingTestWrapper>
    );

    expect(screen.getByText(/explore food families/i)).toBeVisible();

    user.click(screen.getByTestId(category.name));

    waitFor(() => {
      expect(screen.findByTestId("menutitle-title")).toHaveTextContent(category.name);
      expect(window.location.pathname).toBe(`/NutritionalInformation/${category.name}`);
    });
  });
});
