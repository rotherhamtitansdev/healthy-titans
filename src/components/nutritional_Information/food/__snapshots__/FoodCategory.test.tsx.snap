// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Food Category Component renders with no category 1`] = `
<DocumentFragment>
  <div>
    <div
      class="bg-mobileWave sm:bg-wave bg-no-repeat pb-2 bg-span"
    >
      <div
        class="flex flex-wrap h-[22rem]"
      >
        <div
          class="mx-auto pt-[3rem] md:pt-[2rem]"
        >
          <h1
            class="text-titansDarkBlue font-bold text-center sm:mt-[1rem] text-3xl sm:text-5xl lg:text-6xl mt-8 pb-4"
            data-testid="header-title"
          >
            Food & Nutrition
          </h1>
          <div
            class="font-medium text-homepageHeaderSubTitle text-sm md:text-base lg:text-xl text-center w-2/3 lg:w-[45rem] mx-auto"
          >
            Food is important, we all need food to be strong and healthy. From here we can learn about the nutritional values in the foods we eat. This will show you the different types of goodness we can get from the different types of food. We will be able to explore what we need to have a healthy, well-balanced diet including carbohydrates, protein as well as vitamins and minerals.
          </div>
        </div>
      </div>
    </div>
    <div
      class="p-5 md:p-10 pb-20"
    >
      <div
        class="md:px-5"
      >
        <div
          class="font-semibold text-[#243A73] text-2xl md:text-3xl lg:text-4xl py-5"
          data-testid="menutitle-title"
        >
          Explore food families
        </div>
        <div
          class="font-medium text-[#243A73] md:text-xl lg:text-xl md:pb-5"
        >
          Click on a picture to learn more
        </div>
      </div>
      <div
        class="grid grid-cols-2 xs:p-4 lg:grid-cols-auto lg:px-14 xl:px-20 2xl:px-30 3xl:px-36 lg:py-10 gap-[0.5rem] md:gap-y-[1rem] lg:gap-y-[3rem]"
        data-testid="menu-styling"
      >
        <div
          class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg lg:w-[17rem] xl:w-[21rem] xl:h-[21rem] 2xl:w-[25rem] 2xl:h-[25rem] hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
          data-testid="Meat"
          role="button"
          tabindex="0"
        >
          <div
            class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
          >
            <div
              class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
            >
              <span
                class=""
                style="width: 100px; height: 100px; border-radius: 100%; border: 2px solid; border-top-color: #DC476D; border-bottom-color: transparent; border-left-color: #DC476D; border-right-color: #DC476D; display: inline-block; animation: react-spinners-ClipLoader-clip 0.75s 0s infinite linear; animation-fill-mode: both;"
              />
            </div>
            <div
              class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
            >
              Meat
            </div>
          </div>
        </div>
        <div
          class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg lg:w-[17rem] xl:w-[21rem] xl:h-[21rem] 2xl:w-[25rem] 2xl:h-[25rem] hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
          data-testid="Fish"
          role="button"
          tabindex="0"
        >
          <div
            class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
          >
            <div
              class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
            >
              <span
                class=""
                style="width: 100px; height: 100px; border-radius: 100%; border: 2px solid; border-top-color: #DC476D; border-bottom-color: transparent; border-left-color: #DC476D; border-right-color: #DC476D; display: inline-block; animation: react-spinners-ClipLoader-clip 0.75s 0s infinite linear; animation-fill-mode: both;"
              />
            </div>
            <div
              class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
            >
              Fish
            </div>
          </div>
        </div>
        <div
          class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg lg:w-[17rem] xl:w-[21rem] xl:h-[21rem] 2xl:w-[25rem] 2xl:h-[25rem] hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
          data-testid="Fruit & Vegetables"
          role="button"
          tabindex="0"
        >
          <div
            class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
          >
            <div
              class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
            >
              <span
                class=""
                style="width: 100px; height: 100px; border-radius: 100%; border: 2px solid; border-top-color: #DC476D; border-bottom-color: transparent; border-left-color: #DC476D; border-right-color: #DC476D; display: inline-block; animation: react-spinners-ClipLoader-clip 0.75s 0s infinite linear; animation-fill-mode: both;"
              />
            </div>
            <div
              class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
            >
              Fruit & Vegetables
            </div>
          </div>
        </div>
        <div
          class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg lg:w-[17rem] xl:w-[21rem] xl:h-[21rem] 2xl:w-[25rem] 2xl:h-[25rem] hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
          data-testid="Dairy & Eggs"
          role="button"
          tabindex="0"
        >
          <div
            class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
          >
            <div
              class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
            >
              <span
                class=""
                style="width: 100px; height: 100px; border-radius: 100%; border: 2px solid; border-top-color: #DC476D; border-bottom-color: transparent; border-left-color: #DC476D; border-right-color: #DC476D; display: inline-block; animation: react-spinners-ClipLoader-clip 0.75s 0s infinite linear; animation-fill-mode: both;"
              />
            </div>
            <div
              class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
            >
              Dairy & Eggs
            </div>
          </div>
        </div>
        <div
          class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg lg:w-[17rem] xl:w-[21rem] xl:h-[21rem] 2xl:w-[25rem] 2xl:h-[25rem] hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
          data-testid="Fast Food"
          role="button"
          tabindex="0"
        >
          <div
            class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
          >
            <div
              class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
            >
              <span
                class=""
                style="width: 100px; height: 100px; border-radius: 100%; border: 2px solid; border-top-color: #DC476D; border-bottom-color: transparent; border-left-color: #DC476D; border-right-color: #DC476D; display: inline-block; animation: react-spinners-ClipLoader-clip 0.75s 0s infinite linear; animation-fill-mode: both;"
              />
            </div>
            <div
              class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
            >
              Fast Food
            </div>
          </div>
        </div>
        <div
          class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg lg:w-[17rem] xl:w-[21rem] xl:h-[21rem] 2xl:w-[25rem] 2xl:h-[25rem] hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
          data-testid="Sweets"
          role="button"
          tabindex="0"
        >
          <div
            class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
          >
            <div
              class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
            >
              <span
                class=""
                style="width: 100px; height: 100px; border-radius: 100%; border: 2px solid; border-top-color: #DC476D; border-bottom-color: transparent; border-left-color: #DC476D; border-right-color: #DC476D; display: inline-block; animation: react-spinners-ClipLoader-clip 0.75s 0s infinite linear; animation-fill-mode: both;"
              />
            </div>
            <div
              class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
            >
              Sweets
            </div>
          </div>
        </div>
        <div
          class="grid col-span-2 lg:col-span-3 "
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
